src/backend/distributed/transaction/lock_graph.c

```diff
@@ -0,0 +1,772 @@
+		if (!IsSameLockGroup(waitingProc, blockingProc))
+		{
+			BackendData *blockingBackendData = GetBackendDataForProc(blockingProc);
+
+			int lockMode = 1;
+
+			/* check conflicts with every locktype held */
+			for (; lockMode <= numLockModes; lockMode++)
+			{
+				if ((procLock->holdMask & LOCKBIT_ON(lockMode)) &&
+					(conflictMask & LOCKBIT_ON(lockMode)))
+				{
+					bool isBlockingXactWaiting = IsProcessWaitingForLock(blockingProc);
+					if (isBlockingXactWaiting)
+					{
+						remaining->procs[remaining->procCount++] = blockingProc;
+					}
+
+					AddWaitEdge(waitGraph, waitingBackendData, waitingProc,
+								blockingBackendData, blockingProc, isBlockingXactWaiting);
+				}
+			}
+		}
+
+		procLock = (PROCLOCK *) SHMQueueNext(procLocks, &procLock->lockLink,
+											 offsetof(PROCLOCK, lockLink));
+	}
+}
+
+
+/*
+ * AddEdgesForWaitQueue adds an edge to the wait graph for processes in front
+ * of waitingProc in the wait queue that are trying to acquire a conflicting
+ * lock.
+ */
+static void
+AddEdgesForWaitQueue(WaitGraph *waitGraph, PGPROC *waitingProc, PROCStack *remaining)
+{
+	BackendData *waitingBackendData = GetBackendDataForProc(waitingProc);
+
+	/* the lock for which this process is waiting */
+	LOCK *waitLock = waitingProc->waitLock;
+
+	/* determine the conflict mask for the lock level used by the process */
+	LockMethod lockMethodTable = GetLocksMethodTable(waitLock);
+	int conflictMask = lockMethodTable->conflictTab[waitingProc->waitLockMode];
+
+	/* iterate through the wait queue */
+	PROC_QUEUE *waitQueue = &(waitLock->waitProcs);
+	int queueSize = waitQueue->size;
+	PGPROC *currentProc = (PGPROC *) waitQueue->links.next;
+
+	while (queueSize-- > 0 && currentProc != waitingProc)
+	{
+		/* skip processes from the same lock group and ones that don't conflict */
+		if (!IsSameLockGroup(waitingProc, currentProc) &&
+			(LOCKBIT_ON(currentProc->waitLockMode) & conflictMask) != 0)
+		{
+			BackendData *currentBackendData = GetBackendDataForProc(currentProc);
+
+			bool isBlockingXactWaiting = IsProcessWaitingForLock(currentProc);
+			if (isBlockingXactWaiting)
+			{
+				remaining->procs[remaining->procCount++] = currentProc;
+			}
+
+			AddWaitEdge(waitGraph, waitingBackendData, waitingProc, currentBackendData,
+						currentProc, isBlockingXactWaiting);
+		}
+
+		currentProc = (PGPROC *) currentProc->links.next;
+	}
+}
+
+
+/*
+ * AddWaitEdge adds a new wait edge to a wait graph. The nodes in the graph are
+ * transactions and an edge indicates the "waiting" transactions is blocked on
+ * a lock held by the "blocking" transaction.
+ */
+static void
+AddWaitEdge(WaitGraph *waitGraph, BackendData *waitingXactData,
+			PGPROC *waitingProc, BackendData *blockingXactData,
+			PGPROC *blockingProc, bool isBlockingProcessWaiting)
+{
+	WaitEdge *curEdge = AllocWaitEdge(waitGraph);
+
+	curEdge->waitingPid = waitingProc->pid;
+
+	if (IsInDistributedTransaction(waitingXactData))
```

Comment:
Hmm, when a transaction is prepared, we `unset` the backend data. In that case, wouldn't we see a distributed transaction as a local one in the graph? Would that cause any problems?